socket_wrapper.cpp

```diff
@@ -30,43 +34,80 @@ void SocketWrapper::InitSocketAddress(int family, int address, int port)
     m_socket_address.sin_port = port;
 }
 
-void SocketWrapper::SetInAddressWithStr(const char* ip_address_str)
+void Socket::SetInAddressWithStr(const char* ip_address_str)
 {
     if((inet_pton(AF_INET, ip_address_str, &m_socket_address.sin_addr)) <= 0)
-        throw socket_exception("inet_pton returned error");
+        throw SocketException("inet_pton returned error");
 }
 
-void SocketWrapper::Bind()
+void Socket::Bind(const unsigned address, const unsigned port)
 {
+    m_socket_address.sin_addr.s_addr = address;
+    m_socket_address.sin_port = port;
+
     if(bind(m_socket_handle, (const sockaddr*)&m_socket_address, sizeof(m_socket_address)) < 0)
-        throw socket_exception("bind returned error");
+        throw SocketException("bind returned error");
 }
 
-void SocketWrapper::Listen(int flag)
+void Socket::Listen(int backlog)
 {
-    if(listen(m_socket_handle, flag) < 0)
-        throw socket_exception("listen returned error");
+    if(listen(m_socket_handle, backlog) < 0)
+        throw SocketException("listen returned error");
 }
 
-std::unique_ptr<SocketWrapper> SocketWrapper::Accept()
+Socket_sptr Socket::Accept()
 {
     struct sockaddr_in client_address;
 	socklen_t client_address_len = 0;
     int client_socket_handle = accept(m_socket_handle, (sockaddr*)&client_address, &client_address_len);
-    if(client_socket_handle == -1) throw socket_exception("accept returned error");
-    return std::make_unique<SocketWrapper>(client_socket_handle, (sockaddr*)&client_address);
+    if(client_socket_handle == -1) throw SocketException("accept returned error");
+    return std::make_shared<Socket>(client_socket_handle, (sockaddr*)&client_address);
 }
 
-void SocketWrapper::Close()
+void Socket::Close()
 {
     if(close(m_socket_handle) < 0)
-        throw socket_exception("close returned error");
+        throw SocketException("close returned error");
+}
+
+int Socket::Send(const void* buffer, size_t buffer_length, int flags)
+{
+    int retVal = send(m_socket_handle, buffer, buffer_length, flags);
+    if(retVal < 0)
+       throw SocketException("send returned error");
+
+    return retVal;
 }
 
-void SocketWrapper::Send(const void* buffer, size_t buffer_length, int flags)
+int Socket::Recv(void* buffer, size_t buffer_length, int flags)
 {
-    if(send(m_socket_handle, buffer, buffer_length, flags) < 0)
-       throw socket_exception("send returned error");
+    int retVal = recv(m_socket_handle, buffer, buffer_length, flags);
+    if(retVal < 0)
+        throw SocketException("recv returned error");
+
+    return retVal;
 }
 
+int Socket::Read(void* buffer, size_t buffer_length)
+{
+    int retVal = read(m_socket_handle, buffer, buffer_length);
+    if(retVal < 0)
+        throw SocketException("read returned error");
 
+    return retVal;
+}
+
+void Socket::Write(const void* buffer, size_t buffer_length)
```

Comment:
Same as `Send`, `Write` returns number of bytes actually written. We might need them too.